File: crates/fey_math/src/vec2.rs
use crate::{Float, Num, Signed, Vec3, impl_vec, vec3};
use dpi::{LogicalPosition, LogicalSize, PhysicalPosition, PhysicalSize};
use std::fmt::{Display, Formatter};

pub type Vec2F = Vec2<f32>;
pub type Vec2I = Vec2<i32>;
pub type Vec2U = Vec2<u32>;

/// A 2-dimensional vector.
#[repr(C)]
#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Vec2<T> {
    pub x: T,
    pub y: T,
}

/// Create a [`Vec2`].
#[inline]
pub const fn vec2<T>(x: T, y: T) -> Vec2<T> {
    Vec2 { x, y }
}

impl_vec!(
    NAME = Vec2
    SHORT = vec2
    LEN = 2
    FIELDS = (x, y)
    TUPLE = (T, T)
);

impl<T> Vec2<T> {
    /// Swizzle components.
    #[inline]
    pub fn yx(self) -> Self {
        vec2(self.y, self.x)
    }

    /// Return the vector with the x-value replaced.
    #[inline]
    pub fn with_x(self, x: T) -> Self {
        Self { x, ..self }
    }

    /// Return the vector with the y-value replaced.
    #[inline]
    pub fn with_y(self, y: T) -> Self {
        Self { y, ..self }
    }

    /// Add a third dimension to this vector.
    #[inline]
    pub fn with_z(self, z: T) -> Vec3<T> {
        vec3(self.x, self.y, z)
    }
}

impl<T: Num> Vec2<T> {
    /// A normalized vector pointing right, equal to `(1, 0)`.
    pub const RIGHT: Self = vec2(T::ONE, T::ZERO);

    /// A normalized vector pointing down, equal to `(0, 1)`.
    pub const DOWN: Self = vec2(T::ZERO, T::ONE);

    /// A normalized vector pointing south, equal to `(0, 1)`.
    pub const SOUTH: Self = vec2(T::ZERO, T::ONE);

    /// A normalized vector pointing east, equal to `(1, 0)`.
    pub const EAST: Self = vec2(T::ONE, T::ZERO);

    /// A unit vector representing the x-axis, equal to `(1, 0)`.
    pub const X_AXIS: Self = vec2(T::ONE, T::ZERO);

    /// A unit vector representing the y-axis, equal to `(0, 1)`.
    pub const Y_AXIS: Self = vec2(T::ZERO, T::ONE);

    /// Returns the cross-product of this vector and another,
    /// equal to `(x1×y2-y1×x2)`.
    #[inline]
    pub fn cross(self, other: Self) -> T {
        self.x * other.y - self.y * other.x
    }

    /// Given the triangle `(a, b, c)`, and the interpolation values
    /// `ab` and `bc`, returns a barycentric coordinate.
    #[inline]
    pub fn barycentric(a: Self, b: Self, c: Self, t1: T, t2: T) -> Self {
        #[inline]
        fn coord<T: Num>(a: T, b: T, c: T, ab: T, bc: T) -> T {
            a + (b - a) * ab + (c - a) * bc
        }
        vec2(coord(a.x, b.x, c.x, t1, t2), coord(a.y, b.y, c.y, t1, t2))
    }
}

impl<T: Signed> Vec2<T> {
    pub const CARDINAL_DIRS: [Self; 4] = [Self::EAST, Self::SOUTH, Self::WEST, Self::NORTH];

    /// A normalized vector pointing left, equal to `(-1, 0)`.
    pub const LEFT: Self = vec2(T::NEG_ONE, T::ZERO);

    /// A normalized vector pointing up, equal to `(0, -1)`.
    pub const UP: Self = vec2(T::ZERO, T::NEG_ONE);

    /// A normalized vector pointing west, equal to `(-1, 0)`.
    pub const WEST: Self = vec2(T::NEG_ONE, T::ZERO);

    /// A normalized vector pointing north, equal to `(0, -1)`.
    pub const NORTH: Self = vec2(T::ZERO, T::NEG_ONE);

    /// Rotates the vector 90° clockwise and returns the result.
    #[inline]
    pub fn turn_right(self) -> Self {
        vec2(-self.y, self.x)
    }

    /// Rotates the vector 90° counter-clockwise and returns the result.
    #[inline]
    pub fn turn_left(self) -> Self {
        vec2(self.y, -self.x)
    }
}

impl<T: Float> Vec2<T> {
    pub const OCTAL_DIRS: [Self; 8] = [
        Self::EAST,
        Self::SOUTH_EAST,
        Self::SOUTH,
        Self::SOUTH_WEST,
        Self::WEST,
        Self::NORTH_WEST,
        Self::NORTH,
        Self::NORTH_EAST,
    ];

    /// A normalized vector pointing south-east, equal to `(1/√2, 1/√2)`.
    pub const SOUTH_EAST: Self = vec2(T::ONE_OVER_SQRT_2, T::ONE_OVER_SQRT_2);

    /// A normalized vector pointing south-east, equal to `(-1/√2, 1/√2)`.
    pub const SOUTH_WEST: Self = vec2(T::NEG_ONE_OVER_SQRT_2, T::ONE_OVER_SQRT_2);

    /// A normalized vector pointing south-east, equal to `(-1/√2, -1/√2)`.
    pub const NORTH_WEST: Self = vec2(T::NEG_ONE_OVER_SQRT_2, T::NEG_ONE_OVER_SQRT_2);

    /// A normalized vector pointing south-east, equal to `(1/√2, -1/√2)`.
    pub const NORTH_EAST: Self = vec2(T::ONE_OVER_SQRT_2, T::NEG_ONE_OVER_SQRT_2);
}

impl<T: Display> Display for Vec2<T> {
    #[inline]
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        self.x.fmt(f)?;
        f.write_str(", ")?;
        self.y.fmt(f)
    }
}

impl<T> From<PhysicalPosition<T>> for Vec2<T> {
    #[inline]
    fn from(value: PhysicalPosition<T>) -> Self {
        vec2(value.x, value.y)
    }
}

impl<T> From<PhysicalSize<T>> for Vec2<T> {
    #[inline]
    fn from(value: PhysicalSize<T>) -> Self {
        vec2(value.width, value.height)
    }
}

impl<T> From<LogicalPosition<T>> for Vec2<T> {
    #[inline]
    fn from(value: LogicalPosition<T>) -> Self {
        vec2(value.x, value.y)
    }
}

impl<T> From<LogicalSize<T>> for Vec2<T> {
    #[inline]
    fn from(value: LogicalSize<T>) -> Self {
        vec2(value.width, value.height)
    }
}

File: crates/fey_math/lua/Vec2.lua
---@meta

---A 2-dimensional vector.
---@class (exact) Vec2: Vec2Methods
---@field x number
---@field y number
---@operator add(Vec2): Vec2
---@operator sub(Vec2): Vec2
---@operator mul(Vec2|number): Vec2
---@operator div(Vec2|number): Vec2
---@operator unm: Vec2

---@class Vec2Class : Vec2Methods
---@overload fun(x: number, y: number): Vec2
local module = {}

---@class Vec2Methods
local methods = {}

---`(0, 0)`
---@return Vec2
---@nodiscard
function module.zero() end

---`(1, 1)`
---@return Vec2
---@nodiscard
function module.one() end

---`(1, 0)`
---@return Vec2
---@nodiscard
function module.x_axis() end

---`(0, 1)`
---@return Vec2
---@nodiscard
function module.y_axis() end

---`(1, 0)`
---@return Vec2
---@nodiscard
function module.right() end

---`(-1, 0)`
---@return Vec2
---@nodiscard
function module.left() end

---`(0, 1)`
---@return Vec2
---@nodiscard
function module.down() end

---`(0, -1)`
---@return Vec2
---@nodiscard
function module.up() end

---`(1, 0)`
---@return Vec2
---@nodiscard
function module.east() end

---`(√2, √2)`
---@return Vec2
---@nodiscard 
function module.south_east() end

---`(0, 1)`
---@return Vec2
---@nodiscard 
function module.south() end

---`(-√2, √2)`
---@return Vec2
---@nodiscard 
function module.south_west() end

---`(-1, 0)`
---@return Vec2
---@nodiscard 
function module.west() end

---`(-√2, -√2)`
---@return Vec2
---@nodiscard 
function module.north_west() end

---`(0, -1)`
---@return Vec2
---@nodiscard 
function module.north() end

---`(√2, -√2)`
---@return Vec2
---@nodiscard 
function module.north_east() end

---Create a new vector.
---@param x number
---@param y number
---@return Vec2
---@nodiscard
function module.new(x, y) end

---Create a new vector `(val, val)`.
---@param val number
---@return Vec2
---@nodiscard
function module.splat(val) end

---Given the triangle `(a, b, c)`, and the interpolation values `ab` and `bc`, returns a barycentric coordinate.
---@param a Vec2
---@param b Vec2
---@param c Vec2
---@param ab number
---@param bc number
---@return Vec2
---@nodiscard
function module.barycentric(a, b, c, ab, bc) end

---Returns a temporary copy of this value.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.box_clone(self) end

---Set the vector's coordinates.
---@param self Vec2
---@param x number
---@param y number
function methods.set(self, x, y) end

---Returns a copy of this vector with absolute value of all components.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.abs(self) end

---Returns `true` if the two vectors are approximately equal.
---@param self Vec2
---@param other Vec2
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Returns `true` if the vector is approximately equal to `(0, 0)`.
---@param self Vec3
---@return boolean
---@nodiscard
function methods.approx_zero(self) end

---Rounds the vector's components up to the nearest integer.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.ceil(self) end

---Clamp's the vector's components between the components of `min` and `max`.
---@param self Vec2
---@param min Vec2
---@param max Vec2
---@return Vec2
---@nodiscard
function methods.clamp(self, min, max) end

---Returns the cross product of two vectors.
---@param self Vec2
---@param other Vec2
---@return number
---@nodiscard
function methods.cross(self, other) end

---Returns the normalized direction towards the point.
---@param self Vec2
---@param point Vec2
---@return Vec2
---@nodiscard
function methods.dir_to(self, point) end

---Returns the distance between two points.
---@param self Vec2
---@param point Vec2
---@return Vec2
---@nodiscard
function methods.dist(self, point) end

---Returns the dot product of two vectors.
---@param self Vec2
---@param other Vec2
---@return number
---@nodiscard
function methods.dot(self, other) end

---Rounds the vector's components down to the nearest integer.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.floor(self) end

---Returns `true` if the vector is equal to `(0, 0)`.
---@return boolean
---@nodiscard
function methods.is_zero(self) end

---Length of the vector.
---@param self Vec2
---@return number
---@nodiscard
function methods.len(self) end

---Returns a vector with components set to the largest values in the arguments.
---@param self Vec2
---@param ... Vec2
---@return Vec2
---@nodiscard
function methods.max(self, ...) end

---Returns a vector with components set to the smallest values in the arguments.
---@param self Vec2
---@param ... Vec2
---@return Vec2
---@nodiscard
function methods.min(self, ...) end

---Normalizes the vector.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.norm(self) end

---Reflects the vector off the provided normal.
---@param self Vec2
---@param normal Vec2
---@return Vec2
---@nodiscard
function methods.reflect(self, normal) end

---Rounds the vector's components to the nearest integer.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.round(self) end

---Returns a copy of this vector with all components signed (set to `1`, `-1`, or `0`).
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.sign(self) end

---Returns the square between two points.
---@param self Vec2
---@param point Vec2
---@return Vec2
---@nodiscard
function methods.sqr_dist(self, point) end

---Squared length of the vector.
---@param self Vec2
---@return number
---@nodiscard
function methods.sqr_len(self) end

---Returns a copy of this vector with all components rounded towards zero.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.trunc(self) end

---Rotates the vector 90° counter-clockwise.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.turn_left(self) end

---Rotates the vector 90° clockwise.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.turn_right(self) end

---Changes the length of the vector without changing its direction.
---@param self Vec2
---@param new_len number
---@return number
---@nodiscard
function methods.with_len(self, new_len) end

---Replaces the `x` component of the vector.
---@param self Vec2
---@param x number
---@return Vec2
---@nodiscard
function methods.with_x(self, x) end

---Replaces the `y` component of the vector.
---@param self Vec2
---@param y number
---@return Vec2
---@nodiscard
function methods.with_y(self, y) end

---Add a third dimension to this vector.
---@param self Vec2
---@param z number
---@return Vec3
---@nodiscard
function methods.with_z(self, z) end

---Swizzles the vector.
---@param self Vec2
---@return Vec2
---@nodiscard
function methods.yx(self) end 

return module